crates/polars-plan/src/logical_plan/optimizer/type_coercion/mod.rs

```diff
@@ -356,30 +357,69 @@ impl OptimizationRule for TypeCoercionRule {
                         AExpr::Cast {
                             expr: other_node,
                             data_type: DataType::Categorical(None),
-                            // does not matter
                             strict: false,
                         }
                     },
-                    (dt, DataType::Utf8) => {
-                        polars_bail!(ComputeError: "cannot compare {:?} to {:?} type in 'is_in' operation", dt, type_other)
+                    #[cfg(feature = "dtype-decimal")]
+                    (DataType::Decimal(_, _), _) | (_, DataType::Decimal(_, _)) => {
+                        polars_bail!(InvalidOperation: "`is_in` cannot check for {:?} values in {:?} data", &type_other, &type_left)
+                    },
+                    // can't check for more granular time_unit in less-granular time_unit data,   
+                    // or we'll cast away valid/necessary precision (eg: nanosecs to millisecs)
+                    (DataType::Datetime(lhs_unit, _), DataType::Datetime(rhs_unit, _)) => {
+                        match (lhs_unit, rhs_unit) {
+                            (TimeUnit::Nanoseconds, _) => return Ok(None),
+                            (TimeUnit::Microseconds, TimeUnit::Microseconds | TimeUnit::Milliseconds) => return Ok(None),
+                            (TimeUnit::Milliseconds, TimeUnit::Milliseconds) => return Ok(None),
+                            _ => polars_bail!(InvalidOperation: "`is_in` cannot check for {:?} precision values in {:?} Datetime data", &rhs_unit, &lhs_unit) 
+                        }
+                    },
+                    (DataType::Duration(lhs_unit), DataType::Duration(rhs_unit)) => {
+                        match (lhs_unit, rhs_unit) {
+                            (TimeUnit::Nanoseconds, _) => return Ok(None),
```

Comment:
I think we can make this a little bit more DRY. We can implement `PartialOrd` for `TimeUnit` at the enum defenition? Than we can check `if lhs_unit <= rhs_unit => Ok(None)`.

Reply:
Ahhh, `PartialOrd` is what I was looking for... yes, I didn't like being explicit here 😅
Let me go put on my Rust trousers.

Reply:
Done ;)